src/main/ruby/core/string.rb

```diff
@@ -643,21 +643,32 @@ def upto(stop, exclusive=false)
     stop = StringValue(stop)
 
     if stop.size == 1 && size == 1
+      enc = Encoding.compatible?(self.encoding, stop.encoding)
+
       return self if self > stop
       after_stop = stop.getbyte(0) + (exclusive ? 0 : 1)
       current = getbyte(0)
       until current == after_stop
-        yield current.chr
+        yield current.chr.force_encoding(enc)
         current += 1
       end
     else
       unless stop.size < size
         after_stop = exclusive ? stop : stop.succ
         current = self
 
+        begin
+          # If both the start and end values are strings representing integer values, the encoding must be US-ASCII.
+          Truffle.invoke_primitive(:string_to_inum, self, 10, true)
+          Truffle.invoke_primitive(:string_to_inum, stop, 10, true)
+          enc = Encoding::US_ASCII
+        rescue ArgumentError
+          enc = self.encoding
+        end
```

Comment:
This seems strange logic. Would it make sense to use a regexp or `ascii_only?` or so?

Reply:
It's basically what MRI does, except MRI doesn't throw an exception if it can't convert to a `Fixnum`. It does a `FIXNUM_P` check afterwards.